iceoryx_posh/include/iceoryx_posh/popo/smart_chunk.hpp

```diff
@@ -0,0 +1,200 @@
+struct SmartChunkPrivateData<TransmissionInterface, const T, H>
+{
+    SmartChunkPrivateData(cxx::unique_ptr<const T>&& smartchunkUniquePtr) noexcept;
+
+    SmartChunkPrivateData(SmartChunkPrivateData&& rhs) noexcept = default;
+    SmartChunkPrivateData& operator=(SmartChunkPrivateData&& rhs) noexcept = default;
+
+    SmartChunkPrivateData(const SmartChunkPrivateData&) = delete;
+    SmartChunkPrivateData& operator=(const SmartChunkPrivateData&) = delete;
+
+    cxx::unique_ptr<const T> smartchunkUniquePtr;
+};
+} // namespace internal
+
+///
+/// @brief The SmartChunk class is a mutable abstraction over types which are written to loaned shared memory.
+/// These smartchunks are transferable to the iceoryx system.
+///
+template <typename TransmissionInterface,
+          typename T,
+          typename H = cxx::add_const_conditionally_t<mepoo::NoUserHeader, T>>
+class SmartChunk
+{
+    static_assert(std::is_const<T>::value == std::is_const<H>::value,
+                  "The type `T` and the user-header `H` must be equal in their const qualifier to ensure the same "
+                  "access restrictions for the user-header as for the smartchunk data!");
+
+  protected:
+    /// @brief Helper type to enable the constructor for the producer, i.e. when T has no const qualifier
+    template <typename S, typename TT>
+    using ForProducerOnly = std::enable_if_t<std::is_same<S, TT>::value && !std::is_const<TT>::value, S>;
+
+    /// @brief Helper type to enable the constructor for the consumer, i.e. when T has a const qualifier
+    template <typename S, typename TT>
+    using ForConsumerOnly = std::enable_if_t<std::is_same<S, TT>::value && std::is_const<TT>::value, S>;
+
+    /// @brief Helper type to enable some methods only if a user-header is used
+    template <typename R, typename HH>
+    using HasUserHeader =
+        std::enable_if_t<std::is_same<R, HH>::value && !std::is_same<R, mepoo::NoUserHeader>::value, R>;
+
+  public:
+    /// @brief Constructor for a SmartChunk used by the transmitter
+    /// @tparam S is a dummy template parameter to enable the constructor only for non-const T
+    /// @param smartchunkUniquePtr is a `rvalue` to a `cxx::unique_ptr<T>` with to the data of the encapsulated type T
+    /// @param transmitter is a reference to the transmitter to be able to use the `publish`, `send` and `release`
+    /// methods
+    template <typename S = T, typename = ForProducerOnly<S, T>>
+    SmartChunk(cxx::unique_ptr<T>&& smartchunkUniquePtr, TransmissionInterface& transmitter) noexcept;
+
+    /// @brief Constructor for a SmartChunk used by the consumer
+    /// @tparam S is a dummy template parameter to enable the constructor only for const T
+    /// @param smartchunkUniquePtr is a `rvalue` to a `cxx::unique_ptr<T>` with to the data of the encapsulated type T
+    template <typename S = T, typename = ForConsumerOnly<S, T>>
+    SmartChunk(cxx::unique_ptr<const T>&& smartchunkUniquePtr) noexcept;
```

Comment:
I slightly tend to make the ctors of the `SmartChunk` protected in order to be only called by `Sample`, `Request` and `Response`. It can later still be made public if we have a use case for this.

Reply:
The `const` in `cxx::unique_ptr<const T>` should not be necessary since this ctor should already only be enabled if `T` has the const qualifier.

Reply:
> The `const` in `cxx::unique_ptr<const T>` should not be necessary since this ctor should already only be enabled if `T` has the const qualifier.

Intended to make things explicit here. But anyways, will remove it